Add a fullscreen button to the bottom box

eqrion opened the pull request:
[Production](https://profiler.firefox.com/public/02zdxmc6747mp8pa213g9fzjm4fmkgvx2sskrc0/flame-graph/?globalTrackOrder=ylzcy8ybx9zjyazbygy3jy7ysxuzizvxevmx0xgz8yj1A0ytA3A1yezdzgxqA2x7uxj02wiklnwtx1wx6x8xawxdxfxhxixkwxpxrwxtxvwy2y4wy6y9ycydyfyhyiykymwyryuwz7z9zazezfzhzkwzuA4wA6&hiddenGlobalTracks=0wxsxuwA6&hiddenLocalTracksByPid=368-0w5~44511-013wg~46196-0~569-0~783-0w5~788-0~843-0~848-0w2~9345-0w2&profileName=reliost%20with%20download%2C%20running%20on%20Hetzner%20CPX11%2C%202025-09-06&range=74508m22397~75577m7948&sourceView=git%3Agithub.com%2Fmstange%2Fsamply%3Awholesym%2Fsrc%2Fdownloader.rs%3A5489fcef028e40cb646ca776502a25e2b51afda6&thread=y6&transforms=ff-1500&v=11) | [Deploy preview](https://deploy-preview-5605--perf-html.netlify.app/public/02zdxmc6747mp8pa213g9fzjm4fmkgvx2sskrc0/flame-graph/?globalTrackOrder=ylzcy8ybx9zjyazbygy3jy7ysxuzizvxevmx0xgz8yj1A0ytA3A1yezdzgxqA2x7uxj02wiklnwtx1wx6x8xawxdxfxhxixkwxpxrwxtxvwy2y4wy6y9ycydyfyhyiykymwyryuwz7z9zazezfzhzkwzuA4wA6&hiddenGlobalTracks=0wxsxuwA6&hiddenLocalTracksByPid=368-0w5~44511-013wg~46196-0~569-0~783-0w5~788-0~843-0~848-0w2~9345-0w2&profileName=reliost%20with%20download%2C%20running%20on%20Hetzner%20CPX11%2C%202025-09-06&range=74508m22397~75577m7948&sourceView=git%3Agithub.com%2Fmstange%2Fsamply%3Awholesym%2Fsrc%2Fdownloader.rs%3A5489fcef028e40cb646ca776502a25e2b51afda6&thread=y6&transforms=ff-1500&v=11)

Fixes #5583. I wasn't sure where to get an icon for this, so I just re-used an existing one. I can look for something better if you have a preferred source for them.

<img width="262" height="176" alt="Screenshot 2025-09-05 at 2 18 17 PM" src="https://github.com/user-attachments/assets/34a7f817-7134-4067-a313-c95162b1692d" />

eqrion committed:
Add a fullscreen button to the bottom box

flodolo commented on locales/en-US/app.ftl:
```diff
@@ -1169,6 +1169,14 @@ BottomBox--assembly-code-not-available-title = Assembly code not available
 
+# The toggle button for making the bottom box fullscreen.
+BottomBox--hide-fullscreen =
+    .title = Minimize the source and assembly views to no longer be fullscreen.
```
```suggestion
    .title = Restore view
```

Or `Exit fullscreen`.

* Minimize and restore are very different behaviors
* Tooltips shouldn't have a closing period

flodolo commented on locales/en-US/app.ftl:
```diff
@@ -1169,6 +1169,14 @@ BottomBox--assembly-code-not-available-title = Assembly code not available
+
+# The toggle button for making the bottom box fullscreen.
+BottomBox--show-fullscreen =
+    .title = Expand the source and assembly views to be fullscreen.
```
```suggestion
    .title = Show fullscreen
```

I think there's no need to explain this is for source/asm view, given the context.

flodolo reviewed (changes requested):
Not localization, but if the icon has an arrow going out, it should change for the opposite action (restore).

With that said, I think that icon is used for "this link opens in a new window" (even with Profiler), so that's confusing.

eqrion committed:
Update messages for fullscreen button

eqrion committed:
Use url-handling for fullscreen status

eqrion committed:
Listen for escape key in fullscreen

eqrion committed:
Use better icons for fullscreen toggle button

eqrion commented:
I've updated the patch with new messages, new SVG icons for minimize/maximise (from font awesome, [license here I think](https://fontawesome.com/license/free)), escape closes fullscreen, and using the history back button closes fullscreen.

eqrion commented:
Light review ping @mstange

canova commented on src/components/app/BottomBox.tsx:
```diff
@@ -150,6 +154,25 @@ class BottomBoxImpl extends React.PureComponent<Props> {
 
+  constructor(props: Props) {
+    super(props);
+    this._onKeyDown = this._onKeyDown.bind(this);
```
You won't have to bind this if you convert the _onKeyDown function into an arrow function. Like:
```ts
  _onKeyDown = (event: KeyboardEvent) => {
     if (event.key === 'Escape' && this.props.isFullscreen) {
      this.props.toggleBottomBoxFullscreen();
    }
  };

```

canova commented on src/app-logic/url-handling.ts:
```diff
@@ -358,6 +363,9 @@ export function getQueryStringFromUrlState(urlState: UrlState): string {
       }
+      if (isBottomBoxFullscreen) {
+        query.bottomFullscreen = true;
+      }
```
Maybe move it inside `if (isBottomBoxOpenPerPanel[selectedTab])`?

canova reviewed (commented):
Thanks @eqrion for the PR and sorry for the late review!
I'm looking at the current icon in the deploy preview, and I think it's not looking very uniform with the other icons:
<img width="203" height="184" alt="Screenshot 2025-10-24 at 2 26 28 PM" src="https://github.com/user-attachments/assets/1826919e-87ef-4c35-b9af-c1a97b5629d4" />

 We generally use one from [photon icons](https://icons.design.firefox.com/) when possible. I think using icons from there will fit in our UI better.
For example, I see these for the full screen and exit:
https://icons.design.firefox.com/viewer/#fullscreen

Can you try with these icons and see if they indeed fit better? Otherwise we can try the icorn icons here too (inside 16x16):
https://acorn.firefox.com/latest/desktop/styles/iconography-OJ7DtlvA
(acorn is the new design system and the photon is the old one)


I also noticed a subtle issue. STR:
1. Open the bottom box and make it full screen.
2. Instead of minimizing, close the full screen view directly by clicking the "X".
3. Double click another function to open the bottom box.

I would expect the bottom box to open again in non-minimized version. But currently it opens up full screen. I think it makes sense to reset this `isBottomBoxFullscreen` state when we close the bottom box too. What do you think?

eqrion committed:
Switch to Photon icons and swap order

eqrion committed:
Try out acorn design icon

eqrion committed:
Use arrow function for key event

eqrion committed:
Improve state handling for bottom box

eqrion commented:
I tried out both of the icons, I'm not sure which I prefer. I also moved the icon to be next to close button, as I think it logically goes there.

Here's the photon one:
<img width="521" height="665" alt="photon" src="https://github.com/user-attachments/assets/cb6244ea-f085-4978-b893-c75fa3858109" />

And here's the acorn one:
<img width="248" height="491" alt="acorn" src="https://github.com/user-attachments/assets/c75d06f2-d7ce-4407-ba9e-0ee4c849e0fe" />

The PR here has the acorn one now. Let me know which you prefer and I can rollback to the photon one easily.

I also changed it so that closing the bottom box will remove the fullscreen state. I did this by dispatching two actions in this case because I wasn't sure how to have a single reducer update the state for both in the current redux architecture. Let me know if there's a better way to do it.

eqrion commented:
Friendly review ping @canova

mstange reviewed (approved):
Looks great to me! Thank you!

mstange commented:
@flodolo I believe the changes you requested have been applied, could you approve?

eqrion committed:
Merge branch 'main' into fullscreen

eqrion committed:
Fix merge conflicts

eqrion committed:
Merge branch 'main' into fullscreen

flodolo commented on locales/en-US/app.ftl:
```diff
@@ -1260,6 +1260,14 @@ BottomBox--assembly-code-not-available-title = Assembly code not available
+
+# The toggle button for making the bottom box fullscreen.
+BottomBox--show-fullscreen =
+    .title = Show fullscreen
```
At least in Firefox, the opposite of "Exit fullscreen" is just "Fullscreen". Is there a reason not to use the same approach?

eqrion committed:
Reword fullscreen message

canova committed:
Merge branch 'main' into fullscreen
